File: SDE Sheet/DAY 7 Two Pointer/3_SUM.cpp
/*
Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.

Notice that the solution set must not contain duplicate triplets.

 

Example 1:

Input: nums = [-1,0,1,2,-1,-4]
Output: [[-1,-1,2],[-1,0,1]]
Example 2:

Input: nums = []
Output: []
Example 3:

Input: nums = [0]
Output: []
 

Constraints:

0 <= nums.length <= 3000
-105 <= nums[i] <= 105
*/

//1.brute force 
// time complexity n3+nlogn(insertion)
//space complexity n;

vector<vector<int>> threeSum(vector<int>& a) {
        set<vector<int>> temp;
        int n = a.size();
        for(int i =0;i<n;i++){
            for(int j =i+1;j<n;j++){
                for(int k =j+1;k<n;k++){
                    if(a[i]+a[j]+a[k]==0) if(i!=j!=k){
                        vector<int> arr={a[i],a[j],a[k]};
                        sort(arr.begin(),arr.end());
                        temp.insert(arr);
                    }
                }
            }
        }
        vector<vector<int>> ans(temp.begin(),temp.end());
        return ans;
    }

// 2. using hashmap
//time complexity n2;
//space complexity n
vector<vector<int>> threeSum(vector<int>& a) {
        unordered_map<int,int> temp;
        set<vector<int>> s;
        int n = a.size();
        
        //insertion in hash table
        for(int i =0;i<n;i++)temp[a[i]]++;
        for(int i =0;i<n;i++){
            t1=temp[a[i]];
            if(temp[a[i]]==1)temp.erase(a[i]);
            else temp[a[i]]--;
            for(int j =i+1;j<n;j++){
                t2=temp[a[j]];
                if(temp[a[j]]==1)temp.erase(a[j]);
                else temp[a[j]]--;
                if(i!=j)if(temp.find(-a[i]-a[j])!=temp.end()){
                    vector<int> arr={a[i],a[j],-a[i]-a[j]};
                    sort(arr.begin(),arr.end());
                    s.insert(arr);
                }
                temp[a[j]]++;;
            }
            temp[a[i]]++;
        }
        vector<vector<int>> ans(s.begin(),s.end());
        return ans;
    }
//3. loop and 2 pointer
//time complexity n2;
//space complexity O(1);

int inc(vector<int>& a, int i){
        while(i+1<a.size()&&a[i+1]==a[i])i++; 
        i++;
        return i;
    }
    int red(vector<int>& a, int i){
        while(i-1>=0&&a[i-1]==a[i])i--;
        i--;
        return i;
    }
    vector<vector<int>> threeSum(vector<int>& a) {
        vector<vector<int>> ans;
        if(a.size()<3)return ans;
        sort(a.begin(),a.end());   
        
        
        int i =0;
        while(i<a.size()-2){
            int target=-a[i];
            int l=i+1;
            int r=a.size()-1;
            while(l<r){
                if(a[l]+a[r]==target){
                    ans.push_back({a[i],a[l],a[r]});
                    l=inc(a,l);
                    r=red(a,r);
                    }
                else if(a[l]+a[r]>target) r=red(a,r);
                else l=inc(a,l);
            }
            i=inc(a,i);
        }
        return ans;
    }

File: Parito Problem Set/Two Pointers/3sum.py
# 15. 3Sum
# Solved
# Medium
# Topics
# Companies
# Hint
# Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.

# Notice that the solution set must not contain duplicate triplets.

 

# Example 1:

# Input: nums = [-1,0,1,2,-1,-4]
# Output: [[-1,-1,2],[-1,0,1]]
# Explanation: 
# nums[0] + nums[1] + nums[2] = (-1) + 0 + 1 = 0.
# nums[1] + nums[2] + nums[4] = 0 + 1 + (-1) = 0.
# nums[0] + nums[3] + nums[4] = (-1) + 2 + (-1) = 0.
# The distinct triplets are [-1,0,1] and [-1,-1,2].
# Notice that the order of the output and the order of the triplets does not matter.
# Example 2:

# Input: nums = [0,1,1]
# Output: []
# Explanation: The only possible triplet does not sum up to 0.
# Example 3:

# Input: nums = [0,0,0]
# Output: [[0,0,0]]
# Explanation: The only possible triplet sums up to 0.
 

# Constraints:

# 3 <= nums.length <= 3000
# -105 <= nums[i] <= 105


# 1. brute force
# TC O(n3)
# SC O(1)


# 2. loop over every element + 2sum
# TC - O(n2) n2 for loop an dnlogn for sort
# SC- O(1)


# optimized n2 skipping duplicates
# TC - O(n2) n2 for loop and nlogn for sort
# SC - O(1) 
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n=len(nums)
        nums=sorted(nums)
        result=[]

        for i,val in enumerate(nums):
            if i>0 and val==nums[i-1]:
                continue
            l,r=i+1,n-1
            while l<r:
                total=val+nums[l]+nums[r]
                if total==0:
                    result.append([val,nums[l],nums[r]])
                    l+=1
                    while nums[l]==nums[l-1] and l<r:
                        l+=1
                elif total<0:
                    l+=1
                else:
                    r-=1

        return result
                    
                
